### PR TITLE
[prism] Add in element manager

### PR DESCRIPTION
**Pull request by lostluck**
Add in the core element manager for Prism. This is the heart of the runner, tracking watermarks for stages, and when stages are able to fire with which elements.

See https://github.com/apache/beam/issues/24789 for more information.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @johannaojeling 
cc: @jrmccluskey 

This is a big PR, but it is fairly well unit tested at least. This is largely how anything for the runner actually gets done. 

Organization wise, stageState should remain below the ElementManager because aside from tracking a given stage's stage, most of it's methods are around the actual bundle readiness for the stage, which I think is easier to understand after reading the element manager handling. What do you think?

I think this might be the 2nd last PR too. The next one requires element manager in, and I don't think the remainder is meaningfully separable. If anything it's only very large because of the test suites that validate everything.

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+		}
+
+		got := ss.minPendingTimestamp()
+		if got != want {
+			t.Errorf("ss.minPendingTimestamp() = %v, want %v", got, want)
+		}
+	})
+}
+
+func TestStageState_getUpstreamWatermark(t *testing.T) {
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go:190]:
```suggestion
func TestStageState_UpstreamWatermark(t *testing.T) {
```

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+	impulse := makeStageState("impulse", nil, nil, []string{"output"})
+	_, up := impulse.UpstreamWatermark()
+	if got, want := up, mtime.MaxTimestamp; got != want {
+		t.Errorf("impulse.getUpstreamWatermark() = %v, want %v", got, want)
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go:194]:
```suggestion
		t.Errorf("impulse.UpstreamWatermark() = %v, want %v", got, want)
```

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+	}
+
+	dofn := makeStageState("dofn", []string{"input"}, nil, []string{"output"})
+	dofn.updateUpstreamWatermark("input", 42)
+
+	_, up = dofn.UpstreamWatermark()
+	if got, want := up, mtime.Time(42); got != want {
+		t.Errorf("dofn.getUpstreamWatermark() = %v, want %v", got, want)
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go:202]:
```suggestion
		t.Errorf("dofn.UpstreamWatermark() = %v, want %v", got, want)
```

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+	}
+
+	flatten := makeStageState("flatten", []string{"a", "b", "c"}, nil, []string{"output"})
+	flatten.updateUpstreamWatermark("a", 50)
+	flatten.updateUpstreamWatermark("b", 42)
+	flatten.updateUpstreamWatermark("c", 101)
+	_, up = flatten.UpstreamWatermark()
+	if got, want := up, mtime.Time(42); got != want {
+		t.Errorf("flatten.getUpstreamWatermark() = %v, want %v", got, want)
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go:211]:
```suggestion
		t.Errorf("flatten.UpstreamWatermark() = %v, want %v", got, want)
```

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+				pcol, up := ss.UpstreamWatermark()
+				t.Errorf("ss.updateWatermarks(%v,%v); ss.input = %v, want %v (upstream %v %v)", test.minPending, test.minStateHold, got, want, pcol, up)
+			}
+			if got, want := ss.output, test.wantOutput; got != want {
+				pcol, up := ss.UpstreamWatermark()
+				t.Errorf("ss.updateWatermarks(%v,%v); ss.output = %v, want %v (upstream %v %v)", test.minPending, test.minStateHold, got, want, pcol, up)
+			}
+			_, up := outStage.UpstreamWatermark()
+			if got, want := up, test.wantDownstream; got != want {
+				t.Errorf("outStage.getUpstreamWatermark() = %v, want %v", got, want)
```
Comment (johannaojeling) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager_test.go:308]:
```suggestion
				t.Errorf("outStage.UpstreamWatermark() = %v, want %v", got, want)
```

**Review by johannaojeling — COMMENTED** [review]
LGTM

**Comment by lostluck** [comment]
Ah! Thank you kindly!

**Pull request merged**